plutus-core/src/Language/PlutusCore/Flat.hs

```diff
@@ -0,0 +1,349 @@
+
+module Language.PlutusCore.Flat ( encode
+                                , decode
+                                ) where
+
+import           Language.PlutusCore.Core
+import           Language.PlutusCore.DeBruijn
+import           Language.PlutusCore.Lexer.Type
+import           Language.PlutusCore.MkPlc      (TyVarDecl (..), VarDecl (..))
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Universe
+
+import           Data.Proxy
+import           Data.Word                      (Word8)
+import           Flat
+import           Flat.Decoder
+import           Flat.Encoder
+
+{- Note [Stable encoding of PLC]
+READ THIS BEFORE TOUCHING ANYTHING IN THIS FILE
+
+We need the encoding of PLC on the blockchain to be *extremely* stable. It *must not* change
+arbitrarily, otherwise we'll be unable to read back old transactions and validate them.
+
+Consequently we don't use the derivable instances of `Flat` for the PLC types that go
+on the chain.
+
+However, the instances in this file *are* constrained by instances for names, type names,
+and annotations. What's to stop the instances for *those* changing, thus changing
+the overall encoding on the chain?
+
+The answer is that what goes on the chain is *always* a `Program TyName Name ()`. The instances
+for `TyName` and `Name` are nailed down here, and the instance for `()` is standard.
+
+However, having this flexibility allows us to encode e.g. PLC with substantial annotations
+(like position information) in situation where the stability is *not* critical, such as
+for testing.
+-}
+
+{- Note [Encoding/decoding constructor tags using Flat]
+Flat saves space when compared to CBOR by allowing tags to use the minimum
+number of bits required for their encoding.
+
+This requires specialised encode/decode functions for each constructor
+that encodes a different number of possibilities. Here is a list of the
+tags and their used/available encoding possibilities.
+
+| Data type       | Function          | Used | Available |
+|-----------------|-------------------|------|-----------|
+| Static builtins | encodeBuiltin     | 22   | 32        |
+| Kinds           | encodeKind        | 2    | 2         |
+| Types           | encodeType        | 7    | 8         |
+| BuiltinNames    | encodeBuiltinName | 2    | 2         |
+| Terms           | encodeTerm        | 10   | 16        |
+
+For format stability we are manually assigning the tag values to the
+constructors (and we do not use a generic algorithm that may change this order).
+
+Flat uses an extra function in its class definition called `size`. Since we want
+to reserve some space for future data constructors and we don't want to have the
+sizes desynchronised from the encoding and decoding functions we have manual
+implementations for them (if they have any constructors reserved for future use).
+-}
+
+{- Note [Don't use catamorphims!]
+We use Codec.Serialise for encoding.  This uses an intermediate type
```

Comment:
This note doesn't apply.

Reply:
Should I remove it from the CBOR encoding as well?

Reply:
The CBOR code used to use `cata` and I added that comment when I removed it because it's a bit of a trap (and one where the problem isn't obvious if you fall into it).  It was mostly a warning for the future: this is a case where functoriality works against you if you try to be too clever.  I don't know if there's any reason to keep it, or whether it also applies to Flat.